zenodo/modules/deposit/views_rest.py

```diff
@@ -53,3 +69,176 @@ def language():
+        return abort(
+            404, 'No metadata exists for the mimetype "{}".'.format(mimetype))
+
+    @validate_metadata_mimetype
+    @pass_record
+    #@check_deposit_permission('update')
+    def put(self, pid, record, **kwargs):
+        """Create or replace metadata."""
+        stream, content_length, content_md5 = \
+            current_files_rest.upload_factory()
+        mimetype = request.headers.get('Content-Type')
+        metadata_bucket = get_or_create_metadata_bucket(record)
+        if record.is_published(): # maybe pass record to deposit var to be easier to read
+            pub_recid, published_record = record.fetch_published()
+        
+        with db.session.begin_nested():
+            if published_record and not published_record['_buckets'].get('metadata'):
+                published_record['_buckets']['metadata'] = str(metadata_bucket.id)
+                published_record.commit()
+                db.session.add(RecordsBuckets(
+                    record_id=published_record.id, bucket_id=metadata_bucket.id
+                ))
+            key = mimetype
+            obj = ObjectVersion.create(metadata_bucket, key, mimetype=mimetype)
+            obj.set_contents(
+                stream, size=content_length, size_limit=metadata_bucket.size_limit)
+        db.session.commit()
+
+        return jsonify({
+            'message': 'Metadata for mimetype "{}" updated.'.format(mimetype)
+        })
+
+    @validate_metadata_mimetype
+    @pass_record
+    #  @check_deposit_permission('update')
+    def delete(self, pid, record, **kwargs):
+        """Delete metadata."""
+        mimetype = request.headers.get('Content-Type')
+        with unlocked_bucket(record.metafiles.bucket): # not sure
+            if mimetype in record.metafiles:
+                del record.metafiles[mimetype]
+        db.session.commit()
+        return jsonify({
+            'message': 'Metadata for mimetype "{}" deleted.'.format(mimetype)})
+
+    @pass_record
+    #  @check_deposit_permission('read')
+    def options(self, pid, record, **kwargs):
+        """Get a list of all metadata."""
+        metadata_files = (f for f in record.metafiles)
+        # TODO: expose file information...
+        return jsonify({
+            f.key: f.file.size
+            for f in metadata_files
+        })
+
+
+class RecordMetadataResource(MethodView):
+    """Record metadata resource."""
+
+    # NOTE: No permissions (allow all), this is considered metadata (not files)
+    @pass_record
+    def get(self, pid, record, **kwargs):
+        """Get metadata."""
+        # Check querystring first, and then "Accept" header
+        mimetype = (request.args.get('mimetype') or
+                    next((m for m, _ in request.accept_mimetypes), None))
+        if mimetype not in _mimetype_whitelist():
+            abort(400, '"{}" is not an acceptable mimetype.'.format(mimetype))
+        if mimetype in record.metafiles:
```

Comment:
If `record.metafiles` is `None`, this will probably fail

Reply:
It is never none since if the bucket does not exist the iterator will return an empty array.